File: LR3/Realization_C++/FullBinaryTree.cpp
#include "FullBinaryTree.h"
#include <iostream>
#include <queue>
#include <cmath>
#include <fstream>
#include <string>

using namespace std;

TreeNode::TreeNode(int value) : data(value), left(nullptr), right(nullptr) {}

FullBinaryTree::FullBinaryTree() : root(nullptr) {}

FullBinaryTree::~FullBinaryTree() {
    clear();
}

void FullBinaryTree::clearRecursive(TreeNode* current) {
    if (current == nullptr) return;
    clearRecursive(current->left);
    clearRecursive(current->right);
    delete current;
}

bool FullBinaryTree::searchRecursive(TreeNode* current, int value) {
    if (current == nullptr) return false;
    if (current->data == value) return true;
    return searchRecursive(current->left, value) || searchRecursive(current->right, value);
}

int FullBinaryTree::countNodes(TreeNode* current) {
    if (current == nullptr) return 0;
    return 1 + countNodes(current->left) + countNodes(current->right);
}

int FullBinaryTree::getHeight(TreeNode* current) {
    if (current == nullptr) return 0;
    int leftHeight = getHeight(current->left);
    int rightHeight = getHeight(current->right);
    return 1 + (leftHeight > rightHeight ? leftHeight : rightHeight);
}

void FullBinaryTree::printInOrderRecursive(TreeNode* current) {
    if (current == nullptr) return;
    printInOrderRecursive(current->left);
    cout << current->data << " ";
    printInOrderRecursive(current->right);
}

void FullBinaryTree::printPreOrderRecursive(TreeNode* current) {
    if (current == nullptr) return;
    cout << current->data << " ";
    printPreOrderRecursive(current->left);
    printPreOrderRecursive(current->right);
}

void FullBinaryTree::printPostOrderRecursive(TreeNode* current) {
    if (current == nullptr) return;
    printPostOrderRecursive(current->left);
    printPostOrderRecursive(current->right);
    cout << current->data << " ";
}

void FullBinaryTree::add(int value) {
    if (root == nullptr) {
        root = new TreeNode(value);
        cout << "Добавлен корень: " << value << endl;
        return;
    }

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* temp = q.front();
        q.pop();

        if (temp->left == nullptr) {
            temp->left = new TreeNode(value);
            cout << "Добавлен элемент: " << value << endl;
            return;
        } else {
            q.push(temp->left);
        }

        if (temp->right == nullptr) {
            temp->right = new TreeNode(value);
            cout << "Добавлен элемент: " << value << endl;
            return;
        } else {
            q.push(temp->right);
        }
    }
}

bool FullBinaryTree::search(int value) {
    bool found = searchRecursive(root, value);
    if (found) {
        cout << "Элемент " << value << " найден!" << endl;
    } else {
        cout << "Элемент " << value << " не найден!" << endl;
    }
    return found;
}

bool FullBinaryTree::isFull() {
    if (root == nullptr) {
        cout << "Дерево пусто!" << endl;
        return true;
    }

    int nodeCount = countNodes(root);
    int height = getHeight(root);
    int expectedNodes = pow(2, height) - 1;

    if (nodeCount == expectedNodes) {
        cout << "Дерево является полным бинарным деревом!" << endl;
        return true;
    } else {
        cout << "Дерево НЕ является полным бинарным деревом." << endl;
        cout << "Узлов: " << nodeCount << ", Ожидается: " << expectedNodes << endl;
        return false;
    }
}

void FullBinaryTree::printInOrder() {
    if (root == nullptr) {
        cout << "Дерево пусто!" << endl;
        return;
    }
    cout << "In-Order обход: ";
    printInOrderRecursive(root);
    cout << endl;
}

void FullBinaryTree::printPreOrder() {
    if (root == nullptr) {
        cout << "Дерево пусто!" << endl;
        return;
    }
    cout << "Pre-Order обход: ";
    printPreOrderRecursive(root);
    cout << endl;
}

void FullBinaryTree::printPostOrder() {
    if (root == nullptr) {
        cout << "Дерево пусто!" << endl;
        return;
    }
    cout << "Post-Order обход: ";
    printPostOrderRecursive(root);
    cout << endl;
}

void FullBinaryTree::printLevelOrder() {
    if (root == nullptr) {
        cout << "Дерево пусто!" << endl;
        return;
    }

    cout << "Level-Order обход: ";
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();
        cout << current->data << " ";

        if (current->left != nullptr) {
            q.push(current->left);
        }
        if (current->right != nullptr) {
            q.push(current->right);
        }
    }
    cout << endl;
}

int FullBinaryTree::getNodeCount() {
    return countNodes(root);
}

int FullBinaryTree::getTreeHeight() {
    return getHeight(root);
}

void FullBinaryTree::clear() {
    clearRecursive(root);
    root = nullptr;
    cout << "Дерево очищено!" << endl;
}

void FullBinaryTree::saveToFile(string filename) {
    ofstream file(filename);
    if (!file.is_open()) {
        cout << "Ошибка открытия файла для записи!" << endl;
        return;
    }

    if (root == nullptr) {
        file.close();
        cout << "Дерево пусто, файл пуст." << endl;
        return;
    }

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        TreeNode* current = q.front();
        q.pop();

        file << current->data << " ";

        if (current->left != nullptr) {
            q.push(current->left);
        }
        if (current->right != nullptr) {
            q.push(current->right);
        }
    }

    file.close();
    cout << "Дерево сохранено в файл: " << filename << endl;
}

void FullBinaryTree::loadFromFile(string filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cout << "Ошибка открытия файла для чтения!" << endl;
        return;
    }

    clear();

    int value;
    while (file >> value) {
        add(value);
    }

    file.close();
    cout << "Дерево загружено из файла: " << filename << endl;
}

// Бинарная сериализация (Level-Order для сохранения структуры дерева)
void FullBinaryTree::saveToBinaryFile(const std::string& filename) {
    std::ofstream file(filename, std::ios::binary);
    if(!file.is_open()) {
        std::cout << "Ошибка открытия бинарного файла для записи!" << std::endl;
        return;
    }

    if(root == nullptr) {
        int empty = 0;
        file.write(reinterpret_cast<const char*>(&empty), sizeof(int));
        file.close();
        std::cout << "Пустое дерево сохранено в бинарный файл: " << filename << std::endl;
        return;
    }

    std::queue<TreeNode*> q;
    q.push(root);
    int nodeCount = getNodeCount();
    
    // Записываем количество узлов
    file.write(reinterpret_cast<const char*>(&nodeCount), sizeof(int));
    
    // Записываем узлы в порядке Level-Order
    while(!q.empty()) {
        TreeNode* current = q.front();
        q.pop();

        file.write(reinterpret_cast<const char*>(&current->data), sizeof(int));

        bool hasLeft = (current->left != nullptr);
        bool hasRight = (current->right != nullptr);
        file.write(reinterpret_cast<const char*>(&hasLeft), sizeof(bool));
        file.write(reinterpret_cast<const char*>(&hasRight), sizeof(bool));

        if(hasLeft) q.push(current->left);
        if(hasRight) q.push(current->right);
    }

    file.close();
    std::cout << "Дерево сохранено в бинарный файл: " << filename << std::endl;
}

void FullBinaryTree::loadFromBinaryFile(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if(!file.is_open()) {
        std::cout << "Ошибка открытия бинарного файла для чтения!" << std::endl;
        return;
    }

    clear();

    int nodeCount;
    file.read(reinterpret_cast<char*>(&nodeCount), sizeof(int));
    
    if(nodeCount == 0) {
        file.close();
        std::cout << "Пустое дерево загружено из бинарного файла: " << filename << std::endl;
        return;
    }
    
    std::queue<TreeNode*> q;
    int rootValue;
    bool hasLeft, hasRight;
    
    // Читаем корень
    file.read(reinterpret_cast<char*>(&rootValue), sizeof(int));
    file.read(reinterpret_cast<char*>(&hasLeft), sizeof(bool));
    file.read(reinterpret_cast<char*>(&hasRight), sizeof(bool));
    
    root = new TreeNode(rootValue);
    q.push(root);
    nodeCount--;
    
    while(!q.empty() && nodeCount > 0) {
        TreeNode* current = q.front();
        q.pop();
        
        // Читаем левого потомка
        if(hasLeft && nodeCount > 0) {
            int leftValue;
            bool leftHasLeft, leftHasRight;
            file.read(reinterpret_cast<char*>(&leftValue), sizeof(int));
            file.read(reinterpret_cast<char*>(&leftHasLeft), sizeof(bool));
            file.read(reinterpret_cast<char*>(&leftHasRight), sizeof(bool));
            
            current->left = new TreeNode(leftValue);
            q.push(current->left);
            nodeCount--;
        }
        
        // Читаем правого потомка
        if(hasRight && nodeCount > 0) {
            int rightValue;
            bool rightHasLeft, rightHasRight;
            file.read(reinterpret_cast<char*>(&rightValue), sizeof(int));
            file.read(reinterpret_cast<char*>(&rightHasLeft), sizeof(bool));
            file.read(reinterpret_cast<char*>(&rightHasRight), sizeof(bool));
            
            current->right = new TreeNode(rightValue);
            q.push(current->right);
            nodeCount--;
        }
        
        // Читаем флаги для следующего узла
        if(!q.empty() && nodeCount > 0) {
            file.read(reinterpret_cast<char*>(&hasLeft), sizeof(bool));
            file.read(reinterpret_cast<char*>(&hasRight), sizeof(bool));
        }
    }

    file.close();
    std::cout << "Дерево загружено из бинарного файла: " << filename << std::endl;
}

File: Realization_GO/full_binary_tree.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"math"
	"os"
	"strconv"
)

// TreeNode - узел бинарного дерева
type TreeNode struct {
	Data  int
	Left  *TreeNode
	Right *TreeNode
}

// NewTreeNode - конструктор узла
func NewTreeNode(value int) *TreeNode {
	return &TreeNode{
		Data:  value,
		Left:  nil,
		Right: nil,
	}
}

// FullBinaryTree - полное бинарное дерево
type FullBinaryTree struct {
	Root *TreeNode
}

// NewFullBinaryTree - конструктор дерева
func NewFullBinaryTree() *FullBinaryTree {
	return &FullBinaryTree{
		Root: nil,
	}
}

// clearRecursive - рекурсивная очистка
func (f *FullBinaryTree) clearRecursive(current *TreeNode) {
	if current == nil {
		return
	}
	f.clearRecursive(current.Left)
	f.clearRecursive(current.Right)
	current.Left = nil
	current.Right = nil
}

// searchRecursive - рекурсивный поиск
func (f *FullBinaryTree) searchRecursive(current *TreeNode, value int) bool {
	if current == nil {
		return false
	}
	if current.Data == value {
		return true
	}
	return f.searchRecursive(current.Left, value) || f.searchRecursive(current.Right, value)
}

// countNodes - подсчет узлов
func (f *FullBinaryTree) countNodes(current *TreeNode) int {
	if current == nil {
		return 0
	}
	return 1 + f.countNodes(current.Left) + f.countNodes(current.Right)
}

// getHeight - получение высоты
func (f *FullBinaryTree) getHeight(current *TreeNode) int {
	if current == nil {
		return 0
	}
	leftHeight := f.getHeight(current.Left)
	rightHeight := f.getHeight(current.Right)
	if leftHeight > rightHeight {
		return 1 + leftHeight
	}
	return 1 + rightHeight
}

// printInOrderRecursive - рекурсивный InOrder обход
func (f *FullBinaryTree) printInOrderRecursive(current *TreeNode) {
	if current == nil {
		return
	}
	f.printInOrderRecursive(current.Left)
	fmt.Printf("%d ", current.Data)
	f.printInOrderRecursive(current.Right)
}

// printPreOrderRecursive - рекурсивный PreOrder обход
func (f *FullBinaryTree) printPreOrderRecursive(current *TreeNode) {
	if current == nil {
		return
	}
	fmt.Printf("%d ", current.Data)
	f.printPreOrderRecursive(current.Left)
	f.printPreOrderRecursive(current.Right)
}

// printPostOrderRecursive - рекурсивный PostOrder обход
func (f *FullBinaryTree) printPostOrderRecursive(current *TreeNode) {
	if current == nil {
		return
	}
	f.printPostOrderRecursive(current.Left)
	f.printPostOrderRecursive(current.Right)
	fmt.Printf("%d ", current.Data)
}

// Add - добавление элемента
func (f *FullBinaryTree) Add(value int) {
	if f.Root == nil {
		f.Root = NewTreeNode(value)
		fmt.Printf("Добавлен корень: %d\n", value)
		return
	}

	queue := []*TreeNode{f.Root}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.Left == nil {
			current.Left = NewTreeNode(value)
			fmt.Printf("Добавлен элемент: %d\n", value)
			return
		} else {
			queue = append(queue, current.Left)
		}

		if current.Right == nil {
			current.Right = NewTreeNode(value)
			fmt.Printf("Добавлен элемент: %d\n", value)
			return
		} else {
			queue = append(queue, current.Right)
		}
	}
}

// Search - поиск элемента
func (f *FullBinaryTree) Search(value int) bool {
	found := f.searchRecursive(f.Root, value)
	if found {
		fmt.Printf("Элемент %d найден!\n", value)
	} else {
		fmt.Printf("Элемент %d не найден!\n", value)
	}
	return found
}

// IsFull - проверка на полное бинарное дерево
func (f *FullBinaryTree) IsFull() bool {
	if f.Root == nil {
		fmt.Println("Дерево пусто!")
		return true
	}

	nodeCount := f.countNodes(f.Root)
	height := f.getHeight(f.Root)
	expectedNodes := int(math.Pow(2, float64(height))) - 1

	if nodeCount == expectedNodes {
		fmt.Println("Дерево является полным бинарным деревом!")
		return true
	} else {
		fmt.Printf("Дерево НЕ является полным бинарным деревом.\n")
		fmt.Printf("Узлов: %d, Ожидается: %d\n", nodeCount, expectedNodes)
		return false
	}
}

// PrintInOrder - InOrder обход
func (f *FullBinaryTree) PrintInOrder() {
	if f.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	fmt.Print("In-Order обход: ")
	f.printInOrderRecursive(f.Root)
	fmt.Println()
}

// PrintPreOrder - PreOrder обход
func (f *FullBinaryTree) PrintPreOrder() {
	if f.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	fmt.Print("Pre-Order обход: ")
	f.printPreOrderRecursive(f.Root)
	fmt.Println()
}

// PrintPostOrder - PostOrder обход
func (f *FullBinaryTree) PrintPostOrder() {
	if f.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	fmt.Print("Post-Order обход: ")
	f.printPostOrderRecursive(f.Root)
	fmt.Println()
}

// PrintLevelOrder - LevelOrder обход
func (f *FullBinaryTree) PrintLevelOrder() {
	if f.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}

	fmt.Print("Level-Order обход: ")
	queue := []*TreeNode{f.Root}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		fmt.Printf("%d ", current.Data)

		if current.Left != nil {
			queue = append(queue, current.Left)
		}
		if current.Right != nil {
			queue = append(queue, current.Right)
		}
	}
	fmt.Println()
}

// GetNodeCount - получение количества узлов
func (f *FullBinaryTree) GetNodeCount() int {
	return f.countNodes(f.Root)
}

// GetTreeHeight - получение высоты дерева
func (f *FullBinaryTree) GetTreeHeight() int {
	return f.getHeight(f.Root)
}

// Clear - очистка дерева
func (f *FullBinaryTree) Clear() {
	f.clearRecursive(f.Root)
	f.Root = nil
	fmt.Println("Дерево очищено!")
}

// SaveToFile - сохранение в текстовый файл
func (f *FullBinaryTree) SaveToFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла для записи: %s\n", filename)
		return err
	}
	defer file.Close()

	if f.Root == nil {
		fmt.Println("Дерево пусто, файл пуст.")
		return nil
	}

	queue := []*TreeNode{f.Root}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		_, err := fmt.Fprintf(file, "%d ", current.Data)
		if err != nil {
			return err
		}

		if current.Left != nil {
			queue = append(queue, current.Left)
		}
		if current.Right != nil {
			queue = append(queue, current.Right)
		}
	}

	fmt.Printf("Дерево сохранено в файл: %s\n", filename)
	return nil
}

// LoadFromFile - загрузка из текстового файла
func (f *FullBinaryTree) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла для чтения: %s\n", filename)
		return err
	}
	defer file.Close()

	f.Clear()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanWords)

	for scanner.Scan() {
		value, err := strconv.Atoi(scanner.Text())
		if err == nil {
			f.Add(value)
		}
	}

	fmt.Printf("Дерево загружено из файла: %s\n", filename)
	return scanner.Err()
}

// SaveToBinaryFile - сохранение в бинарный файл
func (f *FullBinaryTree) SaveToBinaryFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для записи: %s\n", filename)
		return err
	}
	defer file.Close()

	if f.Root == nil {
		empty := int32(0)
		err := binary.Write(file, binary.LittleEndian, empty)
		if err != nil {
			return err
		}
		fmt.Printf("Пустое дерево сохранено в бинарный файл: %s\n", filename)
		return nil
	}

	nodeCount := int32(f.GetNodeCount())
	err = binary.Write(file, binary.LittleEndian, nodeCount)
	if err != nil {
		return err
	}

	queue := []*TreeNode{f.Root}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		// Запись значения узла
		err = binary.Write(file, binary.LittleEndian, int32(current.Data))
		if err != nil {
			return err
		}

		// Флаги наличия потомков
		hasLeft := current.Left != nil
		hasRight := current.Right != nil

		err = binary.Write(file, binary.LittleEndian, hasLeft)
		if err != nil {
			return err
		}
		err = binary.Write(file, binary.LittleEndian, hasRight)
		if err != nil {
			return err
		}

		if hasLeft {
			queue = append(queue, current.Left)
		}
		if hasRight {
			queue = append(queue, current.Right)
		}
	}

	fmt.Printf("Дерево сохранено в бинарный файл: %s\n", filename)
	return nil
}

// LoadFromBinaryFile - загрузка из бинарного файла
func (f *FullBinaryTree) LoadFromBinaryFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для чтения: %s\n", filename)
		return err
	}
	defer file.Close()

	f.Clear()

	var nodeCount int32
	err = binary.Read(file, binary.LittleEndian, &nodeCount)
	if err != nil {
		return err
	}

	if nodeCount == 0 {
		fmt.Printf("Пустое дерево загружено из бинарного файла: %s\n", filename)
		return nil
	}

	// Чтение корня
	var rootValue int32
	var hasLeft, hasRight bool

	err = binary.Read(file, binary.LittleEndian, &rootValue)
	if err != nil {
		return err
	}
	err = binary.Read(file, binary.LittleEndian, &hasLeft)
	if err != nil {
		return err
	}
	err = binary.Read(file, binary.LittleEndian, &hasRight)
	if err != nil {
		return err
	}

	f.Root = NewTreeNode(int(rootValue))
	queue := []*TreeNode{f.Root}
	nodeCount--

	for len(queue) > 0 && nodeCount > 0 {
		current := queue[0]
		queue = queue[1:]

		// Чтение левого потомка
		if hasLeft && nodeCount > 0 {
			var leftValue int32
			var leftHasLeft, leftHasRight bool

			err = binary.Read(file, binary.LittleEndian, &leftValue)
			if err != nil {
				return err
			}
			err = binary.Read(file, binary.LittleEndian, &leftHasLeft)
			if err != nil {
				return err
			}
			err = binary.Read(file, binary.LittleEndian, &leftHasRight)
			if err != nil {
				return err
			}

			current.Left = NewTreeNode(int(leftValue))
			queue = append(queue, current.Left)
			nodeCount--
		}

		// Чтение правого потомка
		if hasRight && nodeCount > 0 {
			var rightValue int32
			var rightHasLeft, rightHasRight bool

			err = binary.Read(file, binary.LittleEndian, &rightValue)
			if err != nil {
				return err
			}
			err = binary.Read(file, binary.LittleEndian, &rightHasLeft)
			if err != nil {
				return err
			}
			err = binary.Read(file, binary.LittleEndian, &rightHasRight)
			if err != nil {
				return err
			}

			current.Right = NewTreeNode(int(rightValue))
			queue = append(queue, current.Right)
			nodeCount--
		}

		// Чтение флагов для следующего узла
		if len(queue) > 0 && nodeCount > 0 {
			err = binary.Read(file, binary.LittleEndian, &hasLeft)
			if err != nil {
				return err
			}
			err = binary.Read(file, binary.LittleEndian, &hasRight)
			if err != nil {
				return err
			}
		}
	}

	fmt.Printf("Дерево загружено из бинарного файла: %s\n", filename)
	return nil
}
